### PR TITLE
[TensorPipe] Always complete futures from thread pool

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* #38934 [TensorPipe] Close and join TP context at shutdown
* #38933 [TensorPipe] Implement join correctly
* **#38930 [TensorPipe] Always complete futures from thread pool**
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging

Any time we mark a future as complete or set an error on it we call its callbacks, which could be arbitrary user functions and could thus be slow or blocking. The safest behavior is to always defer to the loop.

Differential Revision: [D21703017](https://our.internmc.facebook.com/intern/diff/D21703017/)

Differential Revision: [D21703017](https://our.internmc.facebook.com/intern/diff/D21703017)

### CODE REVIEW

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 1]
```diff
@@ -413,22 +419,17 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
...
-                    } else {
-                      markFutureAsComplete(
-                          std::move(futureResponseMessage),
-                          std::move(responseMessage));
-                    }
-                  });
+              if (responseMessage.type() == MessageType::EXCEPTION) {
+                markFutureWithError(
+                    std::move(futureResponseMessage),
+                    std::string(
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:425]: Other part of the code bases uses `c10::str`/`c10::toString`.

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 2]
```diff
@@ -331,6 +331,12 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
   requestMessage.setId(nextMessageID_++);
   pendingResponseMessage[requestMessage.id()] = futureResponseMessage;
 
+  futureResponseMessage->futMsg.addCallback([this]() {
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:334]: I saw there is another `addCallback` in `TensorPipeAgent::respond`, what about that one?

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp (hunk 1 above)
Comment (lw) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:425]: `c10::toString` gave me an error: my impression is that it's only useful to convert some common types to a meaningful string representation. `c10::str` didn't give me an error but I don't think it does what we want: from the code it seems it just appends (`<<`) each argument, which in our case would print the start and end iterator, whereas what we want is copy the string. Did I miss something?

Note that the ProcessGroup agent is also using `std::string` for this:
https://github.com/pytorch/pytorch/blob/b636f5e324f19ebed867d6c3088580a6d6793859/torch/csrc/distributed/rpc/process_group_agent.cpp#L591-L592

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp (hunk 2 above)
Comment (lw) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:334]: We could add a similar check there too, but I think it's less relevant: the future in the `respond` method is created and marked as complete by the RequestCallback, and I think we can trust that it is done correctly. The future we're dealing with here is marked as complete by the agent itself, and it's against errors in the agent that I'm adding this check.

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:425]: I see, thanks for explaining.

**Review by mrshenli — APPROVED** [review]

**Pull request closed without merging**